File: packages/examples/frameworks/mastra/m3_multi_line.ts
/**
 * Example: Mastra + Hymoex M3 (MoE MultiLine).
 *
 * Multiple teams with an integrator routing across them.
 * Each team has a supervisor function coordinating its agents.
 *
 * Requires: npm install @mastra/core
 */

import { Agent, Mastra } from "@mastra/core";

// --- Hymoex M3: Manager -> Integrators -> Expert Manager -> Teams ---

// Finance team agents
const contractsAgent = new Agent({
  name: "contracts",
  instructions: "You are a contracts expert. Draft and review contracts.",
  model: { provider: "OPEN_AI", name: "gpt-4o-mini" },
});

const paymentsAgent = new Agent({
  name: "payments",
  instructions: "You are a payments expert. Process payroll and invoices.",
  model: { provider: "OPEN_AI", name: "gpt-4o-mini" },
});

// HR team agents
const onboardingAgent = new Agent({
  name: "onboarding",
  instructions: "You are an onboarding specialist. Handle new hire processes.",
  model: { provider: "OPEN_AI", name: "gpt-4o-mini" },
});

const complianceAgent = new Agent({
  name: "compliance",
  instructions: "You are a compliance officer. Ensure regulatory compliance.",
  model: { provider: "OPEN_AI", name: "gpt-4o-mini" },
});

const mastra = new Mastra({
  agents: {
    contracts: contractsAgent,
    payments: paymentsAgent,
    onboarding: onboardingAgent,
    compliance: complianceAgent,
  },
});

// Team supervisor functions
async function financeTeam(query: string): Promise<string> {
  const q = query.toLowerCase();
  const agentName = q.includes("contract") ? "contracts" : "payments";
  const agent = mastra.getAgent(agentName);
  const result = await agent.generate(query);
  return `[Finance/${agentName}] ${result.text}`;
}

async function hrTeam(query: string): Promise<string> {
  const q = query.toLowerCase();
  const agentName = q.includes("onboard") ? "onboarding" : "compliance";
  const agent = mastra.getAgent(agentName);
  const result = await agent.generate(query);
  return `[HR/${agentName}] ${result.text}`;
}

// Integrator routes to the right team
async function integrator(query: string): Promise<string> {
  const q = query.toLowerCase();
  if (
    ["contract", "payment", "payroll", "invoice"].some((w) => q.includes(w))
  ) {
    return financeTeam(query);
  }
  return hrTeam(query);
}

async function main() {
  console.log("Mastra + Hymoex M3: Multi-team with integrator routing");
  console.log("Each team function maps to a Hymoex Team Supervisor");
  // const result = await integrator("Set up payroll for the new hire");
  // console.log(result);
}

main();

File: packages/examples/frameworks/crewai/m3_multi_line.py
"""Example: CrewAI + Hymoex M3 (MoE MultiLine).

Multiple teams as separate Crews, coordinated by a manager crew.
Each team has its own supervisor (Crew) and experts (Agents).

Requires: pip install crewai
"""

from crewai import Agent, Crew, Task

import sys
sys.path.insert(0, "../../../../packages/hymoex-python/src")
from hymoex import (
    ExpertManagerSpec, ExpertSpec, IntegratorSpec, ManagerSpec,
    MultiLine, SupervisorSpec, Team, validate_topology,
)


# --- Hymoex M3: Manager -> Integrators -> Expert Manager -> Teams ---

topology = MultiLine(
    manager=ManagerSpec(objective="Enterprise onboarding"),
    integrators=[IntegratorSpec()],
    expert_manager=ExpertManagerSpec(),
    teams=[
        Team(name="finance", supervisor=SupervisorSpec(), experts=[
            ExpertSpec(domain="contracts"),
            ExpertSpec(domain="payments"),
        ]),
        Team(name="hr", supervisor=SupervisorSpec(), experts=[
            ExpertSpec(domain="onboarding"),
            ExpertSpec(domain="compliance"),
        ]),
    ],
)
print(f"Topology: M3, valid: {validate_topology(topology)['valid']}, agents: {topology.agent_count}")


# --- CrewAI Implementation: One Crew per Hymoex Team ---

# Finance team (Hymoex Team with Supervisor)
contracts_agent = Agent(role="Contracts Expert", goal="Draft and review contracts", backstory="Legal specialist.")
payments_agent = Agent(role="Payments Expert", goal="Process payments", backstory="Financial specialist.")

finance_crew = Crew(
    agents=[contracts_agent, payments_agent],
    tasks=[
        Task(description="Review the employment contract", agent=contracts_agent),
        Task(description="Set up payroll", agent=payments_agent),
    ],
    verbose=True,
)

# HR team (Hymoex Team with Supervisor)
onboarding_agent = Agent(role="Onboarding Expert", goal="Onboard new hires", backstory="HR specialist.")
compliance_agent = Agent(role="Compliance Expert", goal="Ensure regulatory compliance", backstory="Compliance officer.")

hr_crew = Crew(
    agents=[onboarding_agent, compliance_agent],
    tasks=[
        Task(description="Complete onboarding checklist", agent=onboarding_agent),
        Task(description="Verify compliance requirements", agent=compliance_agent),
    ],
    verbose=True,
)


# Integrator: routes to the right team
def integrator_route(query: str) -> Crew:
    """Integrator routes to the appropriate team crew."""
    q = query.lower()
    if any(w in q for w in ["contract", "payment", "payroll"]):
        return finance_crew
    return hr_crew


if __name__ == "__main__":
    print("\nCrewAI + Hymoex M3: Multiple team crews with integrator routing")
    print("Each CrewAI Crew maps to a Hymoex Team (Supervisor + Experts)")
    # Uncomment to run:
    # team = integrator_route("Set up payroll for the new hire")
    # result = team.kickoff()
